internal/metamorph/processor_helpers.go

```diff
@@ -22,78 +24,112 @@ func (p *Processor) GetProcessorMapSize() int {
 	return p.responseProcessor.getMapLen()
 }
 
-func updateStatusMap(statusUpdatesMap map[chainhash.Hash]store.UpdateStatus, statusUpdate store.UpdateStatus) {
-	foundStatusUpdate, found := statusUpdatesMap[statusUpdate.Hash]
-
-	if !found || shouldUpdateStatus(statusUpdate, foundStatusUpdate) {
-		if len(statusUpdate.CompetingTxs) > 0 {
-			statusUpdate.CompetingTxs = mergeUnique(statusUpdate.CompetingTxs, foundStatusUpdate.CompetingTxs)
+func (p *Processor) updateStatusMap(statusUpdate store.UpdateStatus) error {
+	currentStatusUpdate, err := p.getTransactionStatus(statusUpdate.Hash)
+	if err != nil {
+		if errors.Is(err, cache.ErrCacheNotFound) {
+			p.logger.Info("updateStatusMap - status record not found, creating new one", slog.String("hash", statusUpdate.Hash.String()))
+			// if record doesn't exist, save new one
+			return p.setTransactionStatus(statusUpdate)
 		}
+		return err
+	}
 
-		statusUpdatesMap[statusUpdate.Hash] = statusUpdate
+	if shouldUpdateCompetingTxs(statusUpdate, *currentStatusUpdate) {
+		p.logger.Info("updateStatusMap - updating competing txs", slog.String("hash", statusUpdate.Hash.String()))
+		currentStatusUpdate.CompetingTxs = mergeUnique(statusUpdate.CompetingTxs, currentStatusUpdate.CompetingTxs)
 	}
-}
 
-func (p *Processor) updateStatusMap(statusUpdate store.UpdateStatus) (map[chainhash.Hash]store.UpdateStatus, error) { //nolint:unused
-	statusUpdatesMap := p.getStatusUpdateMap()
+	if shouldUpdateStatus(statusUpdate, *currentStatusUpdate) {
+		p.logger.Info("updateStatusMap - updating status", slog.String("hash", statusUpdate.Hash.String()), slog.String("old_status", currentStatusUpdate.Status.String()), slog.String("new_status", statusUpdate.Status.String()))
+		currentStatusUpdate.StatusHistory = append(currentStatusUpdate.StatusHistory, store.StatusWithTimestamp{
+			Status:    currentStatusUpdate.Status,
+			Timestamp: p.now(),
```

Comment:
I think this is not the correct place to set the timestamp. The `currentStatusUpdate` should already contain the timestamp from that moment when the original event happened.

The `PeerTxMessage` type has a field `Start` which we can use to capture the timestamp when the event happened and then update the history accordingly

Reply:
Done